rfcs/text/0428-extending-types.md

```diff
@@ -0,0 +1,677 @@
+- RFC PR: [blockprotocol/blockprotocol#428](https://github.com/blockprotocol/blockprotocol/pull/428)
+- Block Protocol Discussion: [blockprotocol/blockprotocol#439](https://github.com/blockprotocol/blockprotocol/discussions/439)
+
+# Summary
+
+[summary]: #summary
+
+As a follow-up to the [Graph type system RFC](./0352-graph-type-system.md), this RFC will describe the behavior of which types in the type system can be extended and duplicated to enhance reusability while giving more control to users (both block authors and users of embedding applications). Extending types can be considered the same as "type inheritance", but there are some important nuances that make these concepts different.
+
+This RFC is not set out to solve duplicating types or "forking", and this will be considered an unresolved problem in this RFC.
+
+# Motivation
+
+[motivation]: #motivation
+
+Reusing public types in the type system comes with the potential disadvantage of not fully conforming to a user's intention. If a user is interested in a public type but needs certain properties to make the type usable for their use case, they would have to recreate the type themself in the current system.
+
+Allowing for types to be extended in the Block Protocol means that a user could still make use of public types when they want to define types for their domain.
+
+This RFC introduces a way for types to be extended in a way where the reusability and sharing aspects of the Block Protocol are maintained.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Type extension can be seen as the concept of adding properties to an existing entity type `Type` by creating a new type `SubType` that has a specific relation to `Type`.
+Using `SubType` in place of `Type` must be possible when extending a type, which means that existing properties and links may _not_ be modified.
+
+For example, an `Employee` entity type can be an extended version of `Person`. This `Employee` type could contain domain-specific properties, making it more concrete while keeping compatibility with `Person`.
+
+If the `Person` entity type contains required properties `Name` and `Age`, the `Employee` entity type would inherit these properties and perhaps add a `Occupation` property. `Employee` would _not_ be able to overwrite any of the properties given by `Person` e.g. it's not possible to turn `Name` into an array or make it optional. This restriction is important as disabling overwriting enables `Employee` instances to be valid `Person` instances i.e. compatibility with `Person` is kept.
+
+The immediate problems that arise from this definition:
+
+- How do we ensure that `Employee` instances can, in fact, be used in place of `Person` entities in practice?
+- Do multiple super-types impose constraints on extending types?
+- How does having `additionalProperties` in existing schemas influence extended types?
+- How do we define extended types within the BP Type System?
```

Comment:
I don't understand this point, wouldn't this boil down to "how do we define extended types?".

If yes, it seems like something that the document should definitely explain, but not something I'd refer to as an "immediate problem" that arose from the definition - so potentially worth omitting.